File: simple-stack-example-basic-fragment/src/main/java/com/zhuinden/navigationexamplefrag/BaseKey.java
package com.zhuinden.navigationexamplefrag;

import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import com.zhuinden.simplestack.navigator.StateKey;
import com.zhuinden.simplestack.navigator.ViewChangeHandler;
import com.zhuinden.simplestack.navigator.changehandlers.FadeViewChangeHandler;

/**
 * Created by Owner on 2017. 06. 29..
 */

public abstract class BaseKey implements Parcelable {
    public String getFragmentTag() {
        return getClass().getName();
    }

    public final BaseFragment newFragment() {
        BaseFragment fragment = createFragment();
        Bundle bundle = fragment.getArguments();
        if (bundle == null) {
            bundle = new Bundle();
        }
        bundle.putParcelable("KEY", this);
        fragment.setArguments(bundle);
        return fragment;
    }

    protected abstract BaseFragment createFragment();
}

File: simple-stack-example-kotlin/src/main/kotlin/com/zhuinden/simplestackexamplekotlin/BaseKey.kt
package com.zhuinden.simplestackexamplekotlin

import android.os.Bundle
import paperparcel.PaperParcelable

/**
 * Created by Owner on 2017.11.13.
 */
abstract class BaseKey : PaperParcelable {
    val fragmentTag: String
        get() = toString()

    fun newFragment(): BaseFragment {
        val fragment = createFragment()
        var bundle: Bundle? = fragment.arguments
        if (bundle == null) {
            bundle = Bundle()
        }
        bundle.putParcelable("KEY", this)
        fragment.arguments = bundle
        return fragment
    }

    protected abstract fun createFragment(): BaseFragment
}
